File: swift/Sources/MoneyHero/MarketData/MarketDataProtocols.swift
import Foundation

public protocol MarketCacheStoring: Sendable {
    func saveMarketQuote(_ envelope: CachedMarketQuote) async throws
    func saveHistoricalPrices(_ envelope: CachedHistoricalPrices) async throws
    func loadMarketQuote(key: String, nowMs: Int) async throws -> CachedMarketQuote?
    func loadHistoricalPrices(key: String, nowMs: Int) async throws -> CachedHistoricalPrices?
}

public protocol MarketDataProviding: Sendable {
    func fetchStockChart(symbol rawSymbol: String) async throws -> YahooHTTPResponse
    func fetchCurrencyChart(displayedCurrency rawCurrency: String) async throws -> YahooHTTPResponse
    func fetchGoldQuoteChart() async throws -> YahooHTTPResponse
    func fetchGoldHistoryChart() async throws -> YahooHTTPResponse
}

public protocol MarketChartParsing: Sendable {
    func parseStockQuoteAndHistory(data: Data, requestedSymbol: String) throws -> (quote: MarketQuote, history: HistoricalPriceSeries)
    func parseCurrencyQuote(data: Data, requestedSymbol: String, displayedCurrency: String, invertForUSDDisplay: Bool) throws -> MarketQuote
    func parseGoldQuoteAndHistory(quoteData: Data, historyData: Data) throws -> (quote: MarketQuote, history: HistoricalPriceSeries)
}

public protocol ClockProviding: Sendable {
    func nowMilliseconds() -> Int
}

public struct SystemClockProvider: ClockProviding {
    public init() {}

    public func nowMilliseconds() -> Int {
        Int(Date().timeIntervalSince1970 * 1000)
    }
}

File: flutter/lib/market_data/market_data_protocols.dart
import "dart:typed_data";

import "../domain/market_cache_models.dart";
import "../domain/market_history_models.dart";
import "../domain/market_quote_models.dart";

abstract interface class MarketCacheStore {
  Future<void> saveMarketQuote(CachedMarketQuote envelope);
  Future<void> saveHistoricalPrices(CachedHistoricalPrices envelope);
  Future<CachedMarketQuote?> loadMarketQuote({required String key, required int nowMs});
  Future<CachedHistoricalPrices?> loadHistoricalPrices({required String key, required int nowMs});
}

abstract interface class MarketDataProviding {
  Future<YahooHttpResponse> fetchStockChart({required String symbol});
  Future<YahooHttpResponse> fetchCurrencyChart({required String displayedCurrency});
  Future<YahooHttpResponse> fetchGoldQuoteChart();
  Future<YahooHttpResponse> fetchGoldHistoryChart();
}

abstract interface class MarketChartParser {
  ({MarketQuote quote, HistoricalPriceSeries history}) parseStockQuoteAndHistory({
    required Uint8List data,
    required String requestedSymbol,
  });
  MarketQuote parseCurrencyQuote({
    required Uint8List data,
    required String requestedSymbol,
    required String displayedCurrency,
    required bool invertForUsdDisplay,
  });
  ({MarketQuote quote, HistoricalPriceSeries history}) parseGoldQuoteAndHistory({
    required Uint8List quoteData,
    required Uint8List historyData,
  });
}

abstract interface class ClockProvider {
  int nowMilliseconds();
}

class SystemClockProvider implements ClockProvider {
  const SystemClockProvider();

  @override
  int nowMilliseconds() => DateTime.now().millisecondsSinceEpoch;
}

class YahooHttpResponse {
  const YahooHttpResponse({
    required this.requestUrl,
    required this.statusCode,
    required this.data,
  });

  final Uri requestUrl;
  final int statusCode;
  final Uint8List data;
}
